core/client/src/main/java/alluxio/client/block/UnderFileSystemBlockInStream.java

```diff
@@ -0,0 +1,236 @@
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block;
+
+import alluxio.client.Locatable;
+import alluxio.client.UnderFileSystemBlockReader;
+import alluxio.client.block.options.LockBlockOptions;
+import alluxio.client.file.FileSystemContext;
+import alluxio.client.file.options.InStreamOptions;
+import alluxio.exception.AlluxioException;
+import alluxio.metrics.MetricsSystem;
+import alluxio.util.network.NetworkAddressUtils;
+import alluxio.wire.LockBlockResult;
+import alluxio.wire.WorkerNetAddress;
+
+import com.codahale.metrics.Counter;
+import com.google.common.io.Closer;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class provides a streaming API to read a UFS block from an Alluxio worker. The data will
+ * be transferred through an Alluxio worker's dataserver to the client.
+ */
+@NotThreadSafe
+public final class UnderFileSystemBlockInStream extends BufferedBlockInStream implements Locatable {
+  /** Used to manage closeable resources. */
+  private final Closer mCloser;
+  /** If set, the block will not be cached into Alluxio because of this read. */
+  private final boolean mNoCache;
+  /** Whether we are reading from a local Alluxio worker or a remote one. */
+  private final boolean mLocal;
+
+  /** Client to communicate with the remote worker. */
+  private final BlockWorkerClient mBlockWorkerClient;
+  /** The file system context which provides block worker clients. */
+  private final FileSystemContext mContext;
+  /** {@link UnderFileSystemBlockReader} for this instance. */
+  private UnderFileSystemBlockReader mReader;
+
+  /**
+   * Creates an instance of {@link UnderFileSystemBlockInStream}. It keeps polling the worker
+   * until the block is cached to Alluxio or it successfully acquires a UFS read token. If
+   * the block is cached to Alluxio, it returns an {@link BufferedBlockInStream} that reads from
+   * Alluxio.
+   *
+   * @param context the file system context
+   * @param ufsPath the UFS path
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param blockStart the start position of the block in the UFS file
+   * @param workerNetAddress the worker network address
+   * @param options the in stream options
+   * @return the {@link UnderFileSystemBlockInStream} instance or the {@link BufferedBlockOutStream}
+   *         that reads from Alluxio directly
+   * @throws IOException if it fails to create {@link UnderFileSystemBlockInStream}
+   */
+  public static BufferedBlockInStream create(FileSystemContext context, String ufsPath,
```

Comment:
Thanks for the suggestion. Your comment is not that accurate since both of the two cases can poll. I have updated the comment by modifying yours. Let me know that is still unclear